catalog/core/catalog-core-migratable/src/main/java/org/codice/ddf/catalog/migratable/impl/MigrationTaskManager.java

```diff
@@ -0,0 +1,188 @@
+     * export directory.
+     *
+     * @throws MigrationException thrown if the setup fails
+     */
+    public void exportSetup() throws MigrationException {
+        fileWriter.init(this.catalogConfig.getExportPath());
+    }
+
+    /**
+     * Must be called after exportSetup and all calls to exportMetacardQuery. This will attempt
+     * to gracefully shutdown the executor service and handle any exceptions that might have
+     * occured in the asynchronous threads.
+     *
+     * If you are using {@link MigrationTaskManager}; that is, you have constructed an instance,
+     * you must call this method to gracefully release resources and shutdown.
+     *
+     * @throws MigrationException thrown if some of the some of the metacards couldn't be exported
+     */
+    public void exportFinish() throws MigrationException {
+        LOGGER.info("Attempting to shutdown catalog export");
+        taskExecutor.shutdown();
+        boolean isGracefulTermination = false;
+        try {
+            isGracefulTermination = taskExecutor.awaitTermination(1L, TimeUnit.MINUTES);
+        } catch (InterruptedException e) {
+            LOGGER.warn("Executor was interrupted during shutdown: " + e.getMessage(), e);
+        } finally {
+            if (!isGracefulTermination) {
+                taskExecutor.shutdownNow();
+            }
+        }
+    }
+
+    /**
+     * Creates a sublist from the given results and a destination file for metacards based on
+     * the {@link CatalogMigratableConfig} object used by this task manager. An asynchronous
+     * task is started for exporting the metacards and is deferred to the appropriate instance
+     * of {@link MigrationFileWriter} for processing.
+     *
+     * @param results          The results of a catalog query that need to be exported.
+     * @param exportGroupCount The group or page number of the query. This value is not monitored
+     *                         or audited for repeats and is strictly for record keeping by naming
+     *                         the resulting export files appropriately.
+     */
+    public void exportMetacardQuery(final List<Result> results, long exportGroupCount) {
+        for (int i = 0; i < results.size(); i += catalogConfig.getExportCardsPerFile()) {
+            final List<Result> fileResults = results.subList(i,
+                    Math.min((i + catalogConfig.getExportCardsPerFile()), results.size()));
+            final File exportFile = catalogConfig.getExportPath()
+                    .resolve(makeFileName(exportGroupCount, i))
+                    .toFile();
+
+            CatalogWriterCallable writerCallable = createWriterCallable(exportFile,
+                    fileResults,
+                    fileWriter);
+
+            Future<Void> task = taskExecutor.submit(writerCallable);
+            taskData.add(task);
+        }
+
+        List<Future<Void>> finishedTasks = new ArrayList<>();
```

Comment:
May be worth looking at this as well (from http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html)

```
Hook methods
This class provides protected overridable beforeExecute(Thread, Runnable) and afterExecute(Runnable, Throwable) methods that are called before and after execution of each task. These can be used to manipulate the execution environment; for example, reinitializing ThreadLocals, gathering statistics, or adding log entries. Additionally, method terminated() can be overridden to perform any special processing that needs to be done once the Executor has fully terminated.
```